xlml/utils/gpu.py

```diff
@@ -304,48 +124,163 @@ def create_resource(
+    network_link = "global/networks/default"
+    # Use the network interface provided in the network_link argument.
+    network_interface = compute_v1.NetworkInterface()
+    network_interface.network = network_link
+
+    if external_access:
+      access = compute_v1.AccessConfig()
+      access.type_ = compute_v1.AccessConfig.Type.ONE_TO_ONE_NAT.name
+      access.name = "External NAT"
+      access.network_tier = access.NetworkTier.PREMIUM.name
+      network_interface.access_configs = [access]
+
+    # Collect information into the Instance object.
+    instance = compute_v1.Instance()
+    instance.network_interfaces = [network_interface]
+    instance.name = instance_name
+    instance.disks = disks
+    if re.match(r"^zones/[a-z\d\-]+/machineTypes/[a-z\d\-]+$", machine_type):
+      instance.machine_type = machine_type
+    else:
+      instance.machine_type = f"zones/{zone}/machineTypes/{machine_type}"
+
+    instance.scheduling = compute_v1.Scheduling()
+    if accelerators:
+      instance.guest_accelerators = accelerators
+      instance.scheduling.on_host_maintenance = (
+          compute_v1.Scheduling.OnHostMaintenance.TERMINATE.name
+      )
+
+    if metadata:
+      instance.metadata = metadata
+
+    if service_account:
+      instance.service_accounts = [service_account]
+
+    if spot:
+      # Set the Spot VM setting
+      instance.scheduling.provisioning_model = (
+          compute_v1.Scheduling.ProvisioningModel.SPOT.name
+      )
+      instance.scheduling.instance_termination_action = instance_termination_action
+
+    if delete_protection:
+      # Set the delete protection bit
+      instance.deletion_protection = True
+
+    # Prepare the request to insert an instance.
+    request = compute_v1.InsertInstanceRequest()
+    request.zone = zone
+    request.project = project_id
+    request.instance_resource = instance
+
+    # Wait for the create operation to complete.
+    logging.info(f"Creating the {instance_name} instance in {zone}...")
+
+    operation = instance_client.insert(request=request)
+    return operation.name
+
+  @task.sensor(poke_interval=60, timeout=1800, mode="reschedule")
+  def wait_for_resource_creation(operation_name: airflow.XComArg):
+    # Retrives the delete opeartion to check the status.
+    client = compute_v1.ZoneOperationsClient()
+    request = compute_v1.GetZoneOperationRequest(
+        operation=operation_name,
+        project=project_id,
+        zone=zone,
+    )
+    operation = client.get(request=request)
+    status = operation.status.name
+    if status in ("RUNNING", "PENDING"):
+      logging.info(f"Resource create status: {status}, {operation.status_message}")
+      return False
+    else:
+      if operation.error:
+        logging.error(
+            f"Error during resource creation: [Code: {operation.http_error_status_code}]: {operation.http_error_message}",
+        )
+        raise operation.exception() or RuntimeError(operation.http_error_message)
```

Comment:
Have you tested this route? wondering the result for `raise operation.exception() or RuntimeError(operation.http_error_message)`

Reply:
Not yet, I haven't seen the case where request was sent but failed the response.

Reply:
If that's the case, how are we confident this statement is correct/appropriate here?

Reply:
This was referred from example like https://github.com/GoogleCloudPlatform/python-docs-samples/blob/a585e56c69a0da7e64286c54f438cfd863a03270/compute/client_library/snippets/instances/stop.py#L63-L70.

Reply:
Thanks Pei.

```
This method will raise the exception received from `operation.exception()`
        or RuntimeError if there is no exception set, but there is an `error_code`
        set for the `operation`.`
```